File: Editor/lib/game_project/project_template.dart
import 'dart:io';
import 'package:xml/xml.dart' as xml;

class ProjectTemplate {
  final Directory _templatesDir = Directory("ProjectTemplates");

  String projectType;
  String projectFile;
  List<String> folders;

  late String projectName;
  late File icon;
  late File screenshot;
  late String projectFilePath;
  late String iconPath;
  late String screenshotPath;

  ProjectTemplate(
      {required this.projectType,
      required this.projectFile,
      required this.folders}) {
    final Directory templateFolderPath =
        Directory('${_templatesDir.path}/$projectType');
    projectName = _getProjectName();
    iconPath = '${templateFolderPath.path}/icon.png';
    screenshotPath = '${templateFolderPath.path}/screenshot.png';
    projectFilePath = '${templateFolderPath.path}/project.lightning';
    icon = File(iconPath);
    screenshot = File(screenshotPath);
  }

  factory ProjectTemplate.fromXMLFile(File file) {
    String content = file.readAsStringSync();
    return ProjectTemplate.fromXML(content);
  }

  factory ProjectTemplate.fromXML(String xmlStr) {
    xml.XmlDocument document = xml.XmlDocument.parse(xmlStr);

    final xml.XmlElement root = document.rootElement;
    final String projectType = (root.getElement('ProjectType')?.innerText)!;
    final String projectFile = (root.getElement('ProjectFile')?.innerText)!;
    final xml.XmlElement foldersNode = (root.getElement("Folders"))!;
    final List<String> folders = <String>[];

    for (final folder in foldersNode.findElements("Folder")) {
      folders.add(folder.innerText);
    }

    return ProjectTemplate(
        projectType: projectType, projectFile: projectFile, folders: folders);
  }

  String _getProjectName() {
    return projectType
        .replaceAllMapped(
            RegExp(r'([A-Z])'), (Match match) => ' ${match.group(0)}')
        .trim();
  }
}

File: Editor/Common/ProjectTemplate.cs
using System.Runtime.Serialization;

namespace Editor.Common
{
    [DataContract(Namespace = "http://schemas.datacontract.org/2004/07/Lightning.Editor")]
    public class ProjectTemplate
    {
        [DataMember]
        public required string ProjectType { get; set; }

        [DataMember]
        public required string ProjectFile { get; set; }

        [DataMember]
        public required List<string> Folders { get; set; }

        public required byte[] Icon { get; set; }

        public required byte[] Screenshot { get; set; }

        public required string IconFilePath { get; set; }

        public required string ScreenshotFilePath { get; set; }

        public required string ProjectFilePath { get; set; }
    }
}
